src/annotator/test/config-test.js

```diff
@@ -1,86 +1,220 @@
+var proxyquire = require('proxyquire');
+
+var fakeSettings = sinon.stub();
+var fakeExtractAnnotationQuery = {};
+
+var config = proxyquire('../config', {
+  '../shared/settings': fakeSettings,
+  './util/extract-annotation-query': fakeExtractAnnotationQuery,
+});
+var sandbox = sinon.sandbox.create();
+
+function fakeWindow() {
+  return {
     document: {
-      querySelector: fakeQuerySelector,
-      querySelectorAll: function (selector) {
-        var match = fakeQuerySelector(selector);
-        return match ? [match] : [];
-      },
+      querySelector: sinon.stub().returns({href: 'LINK_HREF'}),
     },
-    location: {hash: ''},
+    location: {href: 'LOCATION_HREF'},
   };
+}
 
-  beforeEach(function () {
-    fakeScriptConfig = '';
+describe('src.annotator.config', function() {
+  beforeEach('stub console.warn()', function() {
+    sandbox.stub(console, 'warn');
   });
 
-  it('reads the app src from the link tag', function () {
-    var linkEl = document.createElement('link');
-    linkEl.type = 'application/annotator+html';
-    linkEl.href = 'https://test.hypothes.is/app.html';
-    document.head.appendChild(linkEl);
-    assert.deepEqual(config(window), {
-      app: linkEl.href,
+  beforeEach('reset fakeSettings', function() {
+    fakeSettings.reset();
+    fakeSettings.returns({});
+  });
+
+  beforeEach('reset fakeExtractAnnotationQuery', function() {
+    fakeExtractAnnotationQuery.extractAnnotationQuery = sinon.stub();
+  });
+
+  afterEach('reset the sandbox', function() {
+    sandbox.restore();
+  });
+
+  context("when there's an application/annotator+html <link>", function() {
+    var link;
+
+    beforeEach('add an application/annotator+html <link>', function() {
+      link = document.createElement('link');
+      link.type = 'application/annotator+html';
+      link.href = 'http://example.com/link';
+      document.head.appendChild(link);
+    });
+
+    afterEach('tidy up the link', function() {
+      document.head.removeChild(link);
+    });
+
+    it("returns the <link>'s href as options.app", function() {
+      assert.equal(config(window).app, link.href);
     });
-    document.head.removeChild(linkEl);
   });
 
-  it('reads the #annotation query fragment', function () {
-    var fakeWindow = Object.assign({}, fakeWindowBase, {
-      location: {href:'https://foo.com/#annotations:456'},
+  context("when there's no application/annotator+html <link>", function() {
+    it('throws a TypeError', function() {
+      var window_ = fakeWindow();
+      window_.document.querySelector.returns(null);
+
+      assert.throws(function() { config(window_); }, TypeError);
     });
-    assert.deepEqual(config(fakeWindow), {
-      app: 'app.html',
-      annotations: '456',
+  });
+
+  it('gets the JSON settings from the document', function() {
+    var window_ = fakeWindow();
+
+    config(window_);
+
+    assert.calledOnce(fakeSettings);
+    assert.calledWithExactly(fakeSettings, window_.document);
```

Comment:
I'm always writing a `calledOnce` followed by a `calledWithExactly`. I want something equivalent to Python mocks' `assertCalledOnceWith`. I wonder if there's a way for us to add one of our own to Sinon?